File: 99-Continue/168/SumBalancer.java
import java.util.*;

public class SumBalancer {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt();
        sc.nextLine();

        String[] arr = sc.nextLine().split(" ");

        BinaryTree tree = new BinaryTree();
        tree.buildTree(arr);

        System.out.println(tree.sumBalancer() ? "Yes" : "No");

        sc.close();
    }
}

class Node {
    int data;
    Node left, right;

    Node(int data) {
        this.data = data;
    }
}

class BinaryTree {

    Node root;

    Node buildTree(String[] arr) {

        if (arr.length == 0)
            return null;

        root = new Node(Integer.parseInt(arr[0]));

        Queue<Node> queue = new LinkedList<>();
        queue.offer(root);

        int i = 1;

        while (!queue.isEmpty() && i < arr.length) {

            Node current = queue.poll();

            if (i < arr.length && !arr[i].equals("null")) {
                current.left = new Node(Integer.parseInt(arr[i]));
                queue.offer(current.left);
            }
            i++;

            if (i < arr.length && !arr[i].equals("null")) {
                current.right = new Node(Integer.parseInt(arr[i]));
                queue.offer(current.right);
            }
            i++;
        }

        return root;
    }

    boolean sumBalancer() {
        return check(root).isSumTree;
    }

    private Pair check(Node node) {

        if (node == null)
            return new Pair(true, 0);

        if (node.left == null && node.right == null)
            return new Pair(true, node.data);

        Pair left = check(node.left);
        Pair right = check(node.right);

        boolean ok = left.isSumTree &&
                     right.isSumTree &&
                     node.data == left.sum + right.sum;

        return new Pair(ok, node.data + left.sum + right.sum);
    }

    class Pair {
        boolean isSumTree;
        int sum;

        Pair(boolean isSumTree, int sum) {
            this.isSumTree = isSumTree;
            this.sum = sum;
        }
    }
}

File: 99-Continue/168/SumBalancer.py
from collections import deque

class Node:
    def __init__(self, data: int):
        self.data = data
        self.left = self.right = None


class BinaryTree:
    def __init__(self):
        self.root = None

    def buildTree(self, arr: list[str]) -> Node:
        if len(arr) == 0:
            return None

        self.root = Node(int(arr[0]))
        queue = deque([self.root])

        i = 1
        while queue and i < len(arr):
            current = queue.popleft()

            if i < len(arr) and arr[i] != "null":
                current.left = Node(int(arr[i]))
                queue.append(current.left)
            i += 1

            if i < len(arr) and arr[i] != "null":
                current.right = Node(int(arr[i]))
                queue.append(current.right)
            i += 1

        return self.root

    def isSumTree(self, root):
        if root is None:
            return True, 0

        if root.left is None and root.right is None:
            return True, root.data

        left_ok, left_sum = self.isSumTree(root.left)
        right_ok, right_sum = self.isSumTree(root.right)

        if (
            left_ok
            and right_ok
            and root.data == left_sum + right_sum
        ):
            return True, root.data + left_sum + right_sum

        return False, 0

    def sumBalancer(self):
        result, _ = self.isSumTree(self.root)
        return result


def main():
    int(input())
    arr = input().split()

    tree = BinaryTree()
    tree.buildTree(arr)

    print("Yes" if tree.sumBalancer() else "No")


if __name__ == "__main__":
    main()